Ensure cell's replaced content doesn't get the wrong width

Ahmad-S792 opened the pull request:
#### 0b242bd089483bbbd12084ce184b0bf8e4688c54
<pre>
Ensure cell&apos;s replaced content doesn&apos;t get the wrong width

<a href="https://bugs.webkit.org/show_bug.cgi?id=269763">https://bugs.webkit.org/show_bug.cgi?id=269763</a>
<a href="https://rdar.apple.com/123658956">rdar://123658956</a>

Reviewed by Alan Baradlay.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/03d66d9ce1e602df26a9d544d9258a88b9194349">https://chromium.googlesource.com/chromium/src.git/+/03d66d9ce1e602df26a9d544d9258a88b9194349</a>

We need to make the cell&apos;s height unusable for calculating perecent/calc height
against before computing its preferred width, otherwise replaced children may use it
to scale up their intrinsic width.

* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::computePreferredLogicalWidths):
* LayoutTests/fast/table/percent-height-replaced-content-in-cell-expected.txt:
* LayoutTests/fast/table/percent-height-replaced-content-in-cell.html:
* LayoutTests/fast/table/resources/border-image-srgb-color-profile.png:

Canonical link: <a href="https://commits.webkit.org/288316@main">https://commits.webkit.org/288316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5682349831cf6364333ae2ed6e8486a589718138

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2109 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36539 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87577 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2180 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64259 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22013 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75011 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44536 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1450 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29197 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32547 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72742 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29832 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88934 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9751 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72660 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9977 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70825 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71877 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17905 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16017 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15023 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1127 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9704 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15225 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9578 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13044 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11348 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTableCell.cpp:
```diff
@@ -202,7 +203,17 @@ void RenderTableCell::computePreferredLogicalWidths()
     RenderBlockFlow::computePreferredLogicalWidths();
+    if (contentHeight > -1)
+        setOverridingLogicalHeight(*contentHeight);
+
```
```
 LayoutUnit content_height = HasOverrideLogicalContentHeight() ? OverrideLogicalContentHeight() : LayoutUnit(-1);
 if (content_height > -1)
   SetOverrideLogicalContentHeight(LayoutUnit());
 LayoutBlockFlow::ComputePreferredLogicalWidths();
 if (content_height > -1)
   SetOverrideLogicalContentHeight(content_height);

```
Blink seems to use -1 as a magic number, in WebKit this translates to
```
auto overridingLogicalHeight = this->overridingLogicalHeight();
if (overridingLogicalHeight)
  setOverridingLogicalHeight({ });
RenderBlockFlow::computePreferredLogicalWidths();
if (overridingLogicalHeight)
  setOverridingLogicalHeight(*overridingLogicalHeight);

```
though I am not sure if we want a 0 LayoutUnit here or `clearOverridingLogicalHeight`

Ahmad-S792 replied:
Just to add for history - test fails with `clearOverridingLogicalHeight()`.

Ahmad-S792 committed:
Ensure cell's replaced content doesn't get the wrong width

https://bugs.webkit.org/show_bug.cgi?id=269763
rdar://123658956

Reviewed by Alan Baradlay.

Merge: https://chromium.googlesource.com/chromium/src.git/+/03d66d9ce1e602df26a9d544d9258a88b9194349

We need to make the cell's height unusable for calculating perecent/calc height
against before computing its preferred width, otherwise replaced children may use it
to scale up their intrinsic width.

* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::computePreferredLogicalWidths):
* LayoutTests/fast/table/percent-height-replaced-content-in-cell-expected.txt:
* LayoutTests/fast/table/percent-height-replaced-content-in-cell.html:
* LayoutTests/fast/table/resources/border-image-srgb-color-profile.png:

Canonical link: https://commits.webkit.org/288316@main
